Tweak pollTimedOutRPCs thread synchronization

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30355 Tweak pollTimedOutRPCs thread synchronization**
* #31078 Change type of timeoutFutures_ key to time_point instead of duration

- Postpone acquiring lock.
- Release lock earlier.
- Notify conf var after releasing lock.
- Reduce unnecessary scan on future and future timeout maps.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

xush6528 committed:
Tweak pollTimedOutRPCs thread syncrinization

- Make processTimedOutFutures hold lock
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -509,28 +510,26 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
     }
 
-    const auto timedOutFutures = processTimedOutFutures();
+    const auto timedOutFutures = processTimedOutFutures(std::move(lock));
```
Why do we need to pass in the lock into this function? We're already holding the lock when this function is called (see like 488)

xush6528 replied:
I unlock the lock at the end of this function.

mrshenli replied:
I guess the question is why do we pass the lock to `processTimedOutFutures` instead of holding the lock for it in this function? Doing `unlock` and `notify_all` in this function does not seem to introduce any perf penalty but would make the code more readable?

xush6528 replied:
Since we have to acquire lock before testing rpcRunning_, so we don't want to release the lock and re-acquire the lock in the function.

Unless we find a way to not hholding the lock while testing rpcRunning, we can even put all the code about mutating future map and future timeout map into the helper function, `processTimedOutFutures`.

Exactly, notify the cond var as soon as the lock can be released is just for redability.

jjlilley replied:
To chime in, I think the code would be more readable if we didn't pass the lock to the function, but instead manually signal the condvar after, i.e.

```
    const auto timedOutFutures = processTimedOutFutures();
    lock.unlock();
    futureCV_.notify_all();
```

But if others disagree, this is just a weak preference. :) My feeling here is that we want it to be easy to reason about the correctness of the locking here, and std::move()ing the unique_lock is usually best when we have no choice.

xush6528 replied:
@jjlilley 
In the latest version, it's already the same as what you have in mind.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -291,22 +291,22 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
     auto futureStartTime =
         std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::steady_clock::now().time_since_epoch());
+    // Set infinite timeout if specified.
```
Good catch, we don't need to hold the lock for any of this.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -501,6 +501,7 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
 
     if (sleepTime == INFINITE_TIMEOUT) {
       futureTimeoutCV_.wait(lock);
+      continue;
```
Not completely sure whether we should have this or not. One case that we should optimize for is when a new future with a shorter timeout is created, this thread should wake up. With the continue, we will just recompute the new sleep time and go back to sleep. Without it, we'll check in `processTimedOutFutures` whether the first future is already timed out - maybe we should keep this behavior?

xush6528 replied:
Say it takes time X from `enqueueSend` adding a closer end time request with closer end time to the watchdog thread being waken up and running to get a new `sleepTime`.
The shorter timeout(TTL) is time Y.
I would assume time X is at magnitude of us. While users set timeout at magnitude of second (for practical use cases).
So I believe there is gain here.

xush6528 replied:
Never mind. I realized that the below `futureTimeoutCV_.wait_for(lock, sleepTime);` is the critical path.

For most timeout settings, `futureTimeoutCV_.wait(lock);` it used only once at the beginning or a worker has been silent for a long time. 

There is no much gain doing this. 

Reverting the change.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -509,28 +510,26 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
          << " milliseconds and timed out.";
       const auto exceptionMsg = createExceptionResponse(
           Message({}, {}, MessageType::EXCEPTION), ss.str());
+      // Do not hold the lock while marking futures completed, as
```
Although I can see the argument for this comment being here, maybe it should stay above where the lock.unlock() was to explain that?

xush6528 replied:
Actually, this is not a valid comment. when using the future, we don't need to hold the lock. The lock is for protesting the future maps.

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread syncrinization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -509,28 +510,26 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
 
       const int dst = timedOutFuture.dstRank_;
       recvCounts_.increment(dst);
-      futureCV_.notify_all();
```
I kept this notify_all to be in line with what's done when we process received messages. Unsure if removing this here and doing a notify after processing all timed out futures will cause correctness issues.

xush6528 replied:
Should move that up to before `fm->markCompleted(std::move(message));` too.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -555,6 +554,8 @@ const std::vector<ProcessGroupAgent::FutureInfo> ProcessGroupAgent::
       it = futureTimeouts_.erase(it);
     }
   }
+  lock.unlock();
```
Not sure about this pattern - we pass in a lock to `processTimedOutFutures`, which then unlocks it. The caller of processTimedOutFutures is not necessarily aware that the lock has been unlocked right? I think we can keep it as explicit as it was before, and not pass the lock, unless there is a reason to do so.

xush6528 replied:
Notice I used `std::move(lock)`, unique_lock is not copyable, but movable. We only have one lock instance at the same time in `pollTimedOutRPCs`.

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -497,26 +507,46 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       sleepTime = std::max(remainingTime, std::chrono::milliseconds(0));
     }
 
+    bool shouldUpdatesleepTime = true;
```
shouldUpdatesleepTime -> shouldUpdateSleepTime

xush6528 replied:
outdated.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -497,26 +507,46 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
+              return true;
+            }
+            auto earliestEndTimeInMap = futureTimeouts_.begin()->first;
+            return earliestEndTimeInMap < knownEarliestEndTime;
```
This is duplication to the lambda in the if clause. Let's consolidate them.

xush6528 replied:
outdated. does not exist now.

mrshenli reviewed (commented):
This LGTM. Do we know why we hit failures in MacOS test?

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -471,21 +479,40 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
+      if (!rpcRunning_.load()) {
+        return true;
+      }
+      const steady_clock_time_point& minEndTimeInMap =
```
please handle futureTimeouts_.empty() case, since begin()->first will segfault there

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -471,21 +479,40 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       minEndTime = futureTimeouts_.begin()->first;
     }
 
+    const auto& shouldUpdateMinEndTimePredicate = [&, this]() -> bool {
```
- s/const auto&/auto/ on this line - lambda should typically just be auto.

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -243,28 +243,36 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
-        futureTimeoutCV_.notify_one();
+      // future created or it has earlier end time than other futures in the
+      // map.
+      if (futureTimeouts_.begin()->first == endTime &&
```
nice tweak to the condition here!

jjlilley reviewed (approved):
LGTM!

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit e29a7c8:
* **6/6** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e29a7c867a5dd21e033291b5799eb7dfb6e347b8).

### 5 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3911916) pytorch_linux_xenial_py3_clang5_asan_build (1/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72674297))
```
Dec 12 00:32:23   for (int i = 1; i < tensors.size(); ++i) {
Dec 12 00:32:23                   ~ ^ ~~~~~~~~~~~~~~
Dec 12 00:32:23 1 warning generated.
Dec 12 00:32:23 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o
Dec 12 00:32:23 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/py_rref.cpp.o
Dec 12 00:32:23 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/python_functions.cpp.o
Dec 12 00:32:24 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/python_rpc_handler.cpp.o
Dec 12 00:32:24 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/request_callback_impl.cpp.o
Dec 12 00:32:24 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/rref_context.cpp.o
Dec 12 00:32:24 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/rref.cpp.o
Dec 12 00:32:31 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.cpp:32:24: error: non-static data member defined out-of-line
Dec 12 00:32:31     ProcessGroupAgent::kInfiniteTimeoutTimePoint =
Dec 12 00:32:31     ~~~~~~~~~~~~~~~~~~~^
Dec 12 00:32:31 1 error generated.
Dec 12 00:32:31 caffe2/torch/CMakeFiles/torch_python.dir/build.make:2678: recipe for target 'caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o' failed
Dec 12 00:32:31 make[2]: *** [caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o] Error 1
Dec 12 00:32:31 CMakeFiles/Makefile2:11458: recipe for target 'caffe2/torch/CMakeFiles/torch_python.dir/all' failed
Dec 12 00:32:31 make[1]: *** [caffe2/torch/CMakeFiles/torch_python.dir/all] Error 2
Dec 12 00:32:31 make: *** [all] Error 2
Dec 12 00:32:31 Makefile:138: recipe for target 'all' failed
Dec 12 00:32:31 -- Building version 1.4.0a0+e29a7c8
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3911912) pytorch_xla_linux_xenial_py3_6_clang7_build (2/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72674995))
```
Dec 12 00:33:21   for (int i = 1; i < tensors.size(); ++i) {
Dec 12 00:33:21                   ~ ^ ~~~~~~~~~~~~~~
Dec 12 00:33:21 1 warning generated.
Dec 12 00:33:21 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o
Dec 12 00:33:21 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/py_rref.cpp.o
Dec 12 00:33:22 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/python_functions.cpp.o
Dec 12 00:33:22 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/python_rpc_handler.cpp.o
Dec 12 00:33:22 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/request_callback_impl.cpp.o
Dec 12 00:33:22 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/rref_context.cpp.o
Dec 12 00:33:23 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/rref.cpp.o
Dec 12 00:33:29 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.cpp:32:24: error: non-static data member defined out-of-line
Dec 12 00:33:29     ProcessGroupAgent::kInfiniteTimeoutTimePoint =
Dec 12 00:33:29     ~~~~~~~~~~~~~~~~~~~^
Dec 12 00:33:29 1 error generated.
Dec 12 00:33:29 caffe2/torch/CMakeFiles/torch_python.dir/build.make:2678: recipe for target 'caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o' failed
Dec 12 00:33:29 make[2]: *** [caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o] Error 1
Dec 12 00:33:29 CMakeFiles/Makefile2:11690: recipe for target 'caffe2/torch/CMakeFiles/torch_python.dir/all' failed
Dec 12 00:33:29 make[1]: *** [caffe2/torch/CMakeFiles/torch_python.dir/all] Error 2
Dec 12 00:33:29 make: *** [all] Error 2
Dec 12 00:33:29 Makefile:138: recipe for target 'all' failed
Dec 12 00:33:29 -- Building version 1.4.0a0+e29a7c8
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3911930) pytorch_macos_10_13_py3_build (3/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72675121))
```
Dec 11 16:33:33 [2960/2964] Building CXX object modules/observers/CMakeFiles/caffe2_observers.dir/perf_observer.cc.o
Dec 11 16:33:33 ../modules/observers/perf_observer.cc:190:17: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
Dec 11 16:33:33   if (skipIters <= numRuns_ && sampleRate > 0 && rand() % sampleRate == 0) {
Dec 11 16:33:33       ~~~~~~~~~ ^  ~~~~~~~~
Dec 11 16:33:33 ../modules/observers/perf_observer.cc:232:27: warning: comparison of integers of different signs: 'int' and 'std::__1::vector<caffe2::OperatorBase *, std::__1::allocator<caffe2::OperatorBase *> >::size_type' (aka 'unsigned long') [-Wsign-compare]
Dec 11 16:33:33     for (int idx = 0; idx < operators.size(); ++idx) {
Dec 11 16:33:33                       ~~~ ^ ~~~~~~~~~~~~~~~~
Dec 11 16:33:33 2 warnings generated.
Dec 11 16:33:33 [2961/2964] Linking CXX shared library lib/libcaffe2_observers.dylib
Dec 11 16:33:36 [2962/2964] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o
Dec 11 16:33:36 FAILED: caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o 
Dec 11 16:33:36 /Users/distiller/workspace/clang++  -DAT_PARALLEL_OPENMP=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTHP_BUILD_MAIN_LIB -DTH_BLAS_MKL -DUSE_C10D -DUSE_C10D_GLOO -DUSE_DISTRIBUTED -DUSE_NUMPY -D_FILE_OFFSET_BITS=64 -Dtorch_python_EXPORTS -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -Icaffe2/contrib/aten -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -I../torch/.. -I../torch/../aten/src -I../torch/../aten/src/TH -Icaffe2/aten/src -Ithird_party -I../torch/../third_party/gloo -I../torch/../third_party/onnx -I../torch/csrc -I../torch/csrc/api/include -I../torch/lib -I../torch/lib/libshm -I../caffe2/../torch/csrc/api -I../caffe2/../torch/csrc/api/include -I../c10/.. -Ithird_party/ideep/mkl-dnn/include -I../third_party/ideep/mkl-dnn/src/../include -I../torch/lib/libshm/../../../torch/lib -I../torch/lib/c10d/.. -I../torch/lib/c10d/../../.. -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem /Users/distiller/workspace/miniconda3/include -isystem ../third_party/gemmlowp -isystem ../third_party/neon2sse -isystem ../third_party -isystem ../cmake/../third_party/eigen -isystem /Users/distiller/workspace/miniconda3/include/python3.7m -isystem /Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem /opt/rocm/hip/include -isystem /include -isystem ../third_party/ideep/mkl-dnn/include -isystem ../third_party/ideep/include -Wno-deprecated -fvisibility-inlines-hidden -Wno-deprecated-declarations -Xpreprocessor -fopenmp -I/usr/local/include -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-typedef-redefinition -Wno-unknown-warning-option -Wno-unused-private-field -Wno-inconsistent-missing-override -Wno-aligned-allocation-unavailable -Wno-c++14-extensions -Wno-constexpr-not-const -Wno-missing-braces -Qunused-arguments -fcolor-diagnostics -faligned-new -fno-math-errno -fno-trapping-math -Wno-unused-private-field -Wno-missing-braces -Wno-c++14-extensions -Wno-constexpr-not-const -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3  -isysroot /Applications/Xcode-9.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -mmacosx-version-min=10.9 -fPIC   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o -MF caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o.d -o caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o -c ../torch/csrc/distributed/rpc/process_group_agent.cpp
Dec 11 16:33:36 ../torch/csrc/distributed/rpc/process_group_agent.cpp:32:24: error: non-static data member defined out-of-line
Dec 11 16:33:36     ProcessGroupAgent::kInfiniteTimeoutTimePoint =
Dec 11 16:33:36     ~~~~~~~~~~~~~~~~~~~^
Dec 11 16:33:36 1 error generated.
Dec 11 16:33:36 ninja: build stopped: subcommand failed.
Dec 11 16:33:36 Traceback (most recent call last):
Dec 11 16:33:36   File "setup.py", line 755, in <module>
Dec 11 16:33:36     build_deps()
Dec 11 16:33:36   File "setup.py", line 316, in build_deps
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3911917) pytorch_linux_xenial_py3_6_gcc5_4_build (4/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72675609))
```
Dec 12 00:34:40                  from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/message.h:3,
Dec 12 00:34:40                  from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/future_message.h:3,
Dec 12 00:34:40                  from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/python_functions.h:3,
Dec 12 00:34:40                  from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/python_functions.cpp:1:
Dec 12 00:34:40 /var/lib/jenkins/workspace/c10/util/Exception.h:330:13: note: declared here
Dec 12 00:34:40  inline void deprecated_AT_CHECK() {}
Dec 12 00:34:40              ^
Dec 12 00:34:40 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/request_callback_impl.cpp.o
Dec 12 00:34:40 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/rref_context.cpp.o
Dec 12 00:34:40 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/rref.cpp.o
Dec 12 00:34:50 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.cpp:32:24: error: 'const steady_clock_time_point torch::distributed::rpc::ProcessGroupAgent::kInfiniteTimeoutTimePoint' is not a static data member of 'class torch::distributed::rpc::ProcessGroupAgent'
Dec 12 00:34:50      ProcessGroupAgent::kInfiniteTimeoutTimePoint =
Dec 12 00:34:50                         ^
Dec 12 00:34:50 caffe2/torch/CMakeFiles/torch_python.dir/build.make:2678: recipe for target 'caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o' failed
Dec 12 00:34:50 make[2]: *** [caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o] Error 1
Dec 12 00:34:50 CMakeFiles/Makefile2:11393: recipe for target 'caffe2/torch/CMakeFiles/torch_python.dir/all' failed
Dec 12 00:34:50 make[1]: *** [caffe2/torch/CMakeFiles/torch_python.dir/all] Error 2
Dec 12 00:34:50 Makefile:138: recipe for target 'all' failed
Dec 12 00:34:50 make: *** [all] Error 2
Dec 12 00:34:50 -- Building version 1.4.0a0+11bb6e5
Dec 12 00:34:50 cmake -DBUILD_ENVIRONMENT=pytorch-linux-xenial-py3.6-gcc5.4-build -DBUILD_PYTHON=True -DBUILD_TEST=True -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/var/lib/jenkins/workspace/torch -DCMAKE_PREFIX_PATH=/opt/conda/lib/python3.6/site-packages -DNUMPY_INCLUDE_DIR=/opt/conda/lib/python3.6/site-packages/numpy/core/include -DPYTHON_EXECUTABLE=/opt/conda/bin/python -DPYTHON_INCLUDE_DIR=/opt/conda/include/python3.6m -DPYTHON_LIBRARY=/opt/conda/lib/libpython3.6m.so.1.0 -DTORCH_BUILD_VERSION=1.4.0a0+11bb6e5 -DUSE_NUMPY=True -DWERROR=1 /var/lib/jenkins/workspace
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3911928) pytorch_macos_10_13_cuda10_0_cudnn7_py3_build (5/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72676504))
```
Dec 11 16:37:13 [2960/2964] Building CXX object modules/observers/CMakeFiles/caffe2_observers.dir/perf_observer.cc.o
Dec 11 16:37:13 ../modules/observers/perf_observer.cc:190:17: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
Dec 11 16:37:13   if (skipIters <= numRuns_ && sampleRate > 0 && rand() % sampleRate == 0) {
Dec 11 16:37:13       ~~~~~~~~~ ^  ~~~~~~~~
Dec 11 16:37:13 ../modules/observers/perf_observer.cc:232:27: warning: comparison of integers of different signs: 'int' and 'std::__1::vector<caffe2::OperatorBase *, std::__1::allocator<caffe2::OperatorBase *> >::size_type' (aka 'unsigned long') [-Wsign-compare]
Dec 11 16:37:13     for (int idx = 0; idx < operators.size(); ++idx) {
Dec 11 16:37:13                       ~~~ ^ ~~~~~~~~~~~~~~~~
Dec 11 16:37:13 2 warnings generated.
Dec 11 16:37:13 [2961/2964] Linking CXX shared library lib/libcaffe2_observers.dylib
Dec 11 16:37:16 [2962/2964] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o
Dec 11 16:37:16 FAILED: caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o 
Dec 11 16:37:16 /Users/distiller/workspace/clang++  -DAT_PARALLEL_OPENMP=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTHP_BUILD_MAIN_LIB -DTH_BLAS_MKL -DUSE_C10D -DUSE_C10D_GLOO -DUSE_DISTRIBUTED -DUSE_NUMPY -D_FILE_OFFSET_BITS=64 -Dtorch_python_EXPORTS -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -Icaffe2/contrib/aten -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -I../torch/.. -I../torch/../aten/src -I../torch/../aten/src/TH -Icaffe2/aten/src -Ithird_party -I../torch/../third_party/gloo -I../torch/../third_party/onnx -I../torch/csrc -I../torch/csrc/api/include -I../torch/lib -I../torch/lib/libshm -I../caffe2/../torch/csrc/api -I../caffe2/../torch/csrc/api/include -I../c10/.. -Ithird_party/ideep/mkl-dnn/include -I../third_party/ideep/mkl-dnn/src/../include -I../torch/lib/libshm/../../../torch/lib -I../torch/lib/c10d/.. -I../torch/lib/c10d/../../.. -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem /Users/distiller/workspace/miniconda3/include -isystem ../third_party/gemmlowp -isystem ../third_party/neon2sse -isystem ../third_party -isystem ../cmake/../third_party/eigen -isystem /Users/distiller/workspace/miniconda3/include/python3.7m -isystem /Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem /opt/rocm/hip/include -isystem /include -isystem ../third_party/ideep/mkl-dnn/include -isystem ../third_party/ideep/include -Wno-deprecated -fvisibility-inlines-hidden -Wno-deprecated-declarations -Xpreprocessor -fopenmp -I/usr/local/include -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-typedef-redefinition -Wno-unknown-warning-option -Wno-unused-private-field -Wno-inconsistent-missing-override -Wno-aligned-allocation-unavailable -Wno-c++14-extensions -Wno-constexpr-not-const -Wno-missing-braces -Qunused-arguments -fcolor-diagnostics -faligned-new -fno-math-errno -fno-trapping-math -Wno-unused-private-field -Wno-missing-braces -Wno-c++14-extensions -Wno-constexpr-not-const -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3  -isysroot /Applications/Xcode-9.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -mmacosx-version-min=10.9 -fPIC   -DCAFFE2_USE_GLOO -DCUDA_HAS_FP16=1 -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o -MF caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o.d -o caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/rpc/process_group_agent.cpp.o -c ../torch/csrc/distributed/rpc/process_group_agent.cpp
Dec 11 16:37:16 ../torch/csrc/distributed/rpc/process_group_agent.cpp:32:24: error: non-static data member defined out-of-line
Dec 11 16:37:16     ProcessGroupAgent::kInfiniteTimeoutTimePoint =
Dec 11 16:37:16     ~~~~~~~~~~~~~~~~~~~^
Dec 11 16:37:16 1 error generated.
Dec 11 16:37:16 ninja: build stopped: subcommand failed.
Dec 11 16:37:16 Traceback (most recent call last):
Dec 11 16:37:16   File "setup.py", line 755, in <module>
Dec 11 16:37:16     build_deps()
Dec 11 16:37:16   File "setup.py", line 316, in build_deps
```
### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>caffe2_onnx_py3_6_clang7_ubuntu16_04_build</sup> | <sup>Build</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

xush6528 committed:
Update on "Tweak pollTimedOutRPCs thread synchronization"


- Make processTimedOutFutures hold lock.
- Reduce unnecessary scan on future and future timeout maps.
- Reduce the scope of lock at a spot.
- Avoid repeatedly wake up if user set timeout = 0.

Differential Revision: [D5516149](https://our.internmc.facebook.com/intern/diff/D5516149/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -467,11 +473,8 @@ void ProcessGroupAgent::listenLoop() {
 
 void ProcessGroupAgent::pollTimedOutRPCs() {
-  while (true) {
+  while (rpcRunning_.load()) {
```
Don't we need to hold the lock here while checking `rpcRunning`? That's why there was the `while(true)` before

xush6528 replied:
@rohan-varma 

The lock was to avoid `::shutdown()` noitifying when the watch thread is not waiting on the cond var.

With the predicate, we can allow the watch dog to miss the notification.
Even if the watch dog thread misses the notification, the predicate will evaluate to `false` (since `rpcRunning_` has already been set to False), the thread will not wait on the cond var, and gets back to the loop condition and decide it deos not need to loop again.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -483,21 +486,40 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
-    if (!rpcRunning_.load()) {
-      return;
+    if (shouldUpdateMinEndTime) {
+      continue;
```
With this `continue`, I guess this means processing of the timed out futures is going to be skipped until we do another round of updating the `minEndTime`? If there is a timed out future, then there will be some delay in processing it, do we need to have this continue?

xush6528 replied:
@rohan-varma 

Sorry we thought we are closed on https://github.com/pytorch/pytorch/pull/30355#discussion_r349846014.

Two conditions it will stop waiting on the cond var.

1) The cond var is notified, it means there is an ealier `minEndTime` inserted into the map.
So it will go back to the beggining of the loop, update the `minEndTime` and 

    1) if `now >= the latest minEndTime`, it will **process the futures immediately** without waiting on the future.
    2) if `now < the latest minEndTime`, it will wai until `minEndTime`

2) Time is up. `now >= the latest minEndTime`. It will wake from the wait and and process the timed out futures.
